usr/share/rear/lib/serial-functions.sh

```diff
@@ -6,29 +6,36 @@ function get_serial_console_devices () {
-    # cf. "Maintain backward compatibility" in https://github.com/rear/rear/wiki/Coding-Style
-    # Furthermore 'sort' results that /dev/hvsi* devices appear before /dev/ttyS* devices
-    # so the create_grub2_serial_entry function in lib/bootloader-functions.sh
-    # which uses by default the first one and skips the rest will result that
-    # the first /dev/hvsi* device becomes used for the GRUB serial console by default
-    # which looks right because /dev/hvsi* devices should exist only on systems
-    # that have the HVSI driver loaded (a console driver for IBM's p5 servers)
-    # cf. https://lwn.net/Articles/98442/
-    # and it seems right that when special console drivers are loaded
-    # then their devices should be preferred by default:
-    ls /dev/ttyS[0-9]* /dev/hvsi[0-9]* | sort
+    # Scan the kernel command line of the currently running original system
+    # for 'console=<device>[,<options>]' settings e.g. 'console=ttyS1,9600n8 ... console=ttyS3 ... console=tty0'
+    # and extract the specified serial device nodes e.g. ttyS1 -> /dev/ttyS1 ... ttyS3 -> /dev/ttyS3
+    local kernel_option console_option_value console_option_device
+    for kernel_option in $( cat /proc/cmdline ) ; do
+        # Continue with next kernel option when the option name (part before leftmost "=") is not 'console':
+        test "${kernel_option%%=*}" = "console" || continue
+        # Get the console option value (part after leftmost "=") e.g. 'ttyS1,9600n8' 'ttyS3' 'tty0'
+        console_option_value="${kernel_option#*=}"
+        # Get the console option device (part before leftmost optional ',' separator) e.g. 'ttyS1' 'ttyS3' 'tty0'
+        console_option_device="${console_option_value%%,*}"
+        # Continue with next kernel option when the current console option device is no serial device (exclude 'tty0').
+        # The special /dev/hvsi* devices should exist only on systems that have the HVSI driver loaded
+        # (a console driver for IBM's p5 servers) cf. https://lwn.net/Articles/98442/
+        [[ $console_option_device == ttyS* ]] || [[ $console_option_device == hvsi* ]] || continue
+        # Test that the matching serial device node e.g. ttyS1 -> /dev/ttyS1 and ttyS3 -> /dev/ttyS3' exists
+        # to avoid that this automated serial console setup may not work in the ReaR recovery system
+        # when serial device nodes get specified for the recovery system that do not exist
+        # in the currently running original system because the default assumption is
+        # that the replacement system has same hardware as the original system,
+        # cf. https://github.com/rear/rear/pull/2749#issuecomment-1196650631
+        # (if needed the user can specify what he wants via SERIAL_CONSOLE_DEVICES, see above).
+        # Intentionally /dev/$console_option_device is unquoted to let the test also fail
+        # when $console_option_device is not a single non-empty word (then something is wrong):
```

Comment:
If you want to protect against the case where `$console_option_device` is not a single non-empty word, I would avoid doing it this way. The behavior of `test` is governed by the number of arguments and the rules are quite complicated. You don't know how many words will `$console_option_device` expand to.

Reply:
@pcahyna 
what is the right method to check $var a single non-empty word?

Reply:
I would quote the whole thing. How likely is that `test -c` would succeed on something like that? Something would have to be broken on the system if there are device nodes whose names are not single non-empty words.

Reply:
I have it:
In this case it helps that 'test -c' is used
and that results non-zero exit code
when more than one word is provided:
```
# test -c "/dev/ttyS2" && echo y || echo n
y

# test -c "/dev/ttyS2 foo" && echo y || echo n
n

# test -c " " && echo y || echo n
n
```

Reply:
Fixed via
https://github.com/rear/rear/pull/2961/commits/26fdbec25aca11e437c6d909a6aa3e00ebce76e1
and
https://github.com/rear/rear/pull/2961/commits/8319cee5cf3f04ef10772c855d20beaabf05fb0f

@pcahyna 
thank you for your precise review!